vhdx/src/vhdx_io.rs

```diff
@@ -0,0 +1,232 @@
+
+            let bat_entry = match bat.get(sector.bat_index as usize) {
+                Some(entry) => (*entry).bat_entry,
+                None => {
+                    return Err(Error::InvalidBatIndex);
+                }
+            };
+
+            match bat_entry & vhdx_bat::BAT_STATE_BIT_MASK {
+                vhdx_bat::PAYLOAD_BLOCK_NOT_PRESENT
+                | vhdx_bat::PAYLOAD_BLOCK_UNDEFINED
+                | vhdx_bat::PAYLOAD_BLOCK_UNMAPPED
+                | vhdx_bat::PAYLOAD_BLOCK_UNMAPPED_V095
+                | vhdx_bat::PAYLOAD_BLOCK_ZERO => {}
+                vhdx_bat::PAYLOAD_BLOCK_FULLY_PRESENT => {
+                    // should we put lock here or in the calling function?
+                    f.seek(SeekFrom::Start(sector.file_offset))
+                        .map_err(Error::ReadSectorBlock)?;
+                    f.read(
+                        &mut buf[read_count
+                            ..(read_count + (sector.free_sectors * SECTOR_SIZE) as usize)],
+                    )
+                    .map_err(Error::ReadSectorBlock)?;
+                }
+                vhdx_bat::PAYLOAD_BLOCK_PARTIALLY_PRESENT => {
+                    return Err(Error::UnsupportedMode);
+                }
+                _ => {
+                    return Err(Error::InvalidBatEntryState);
+                }
+            };
+            sector_count -= sector.free_sectors;
+            sector_index += sector.free_sectors;
+            read_count = sector.free_bytes as usize;
+        };
+    }
+    Ok(read_count)
+}
+
+/// VHDx IO write routine: requires relative sector index and count for the desired data
+pub fn write(
+    f: &mut File,
+    buf: &[u8],
+    disk_spec: &DiskSpec,
+    bat_offset: u64,
+    bat: &mut Vec<BatEntry>,
+    mut sector_index: u64,
+    mut sector_count: u64,
+) -> Result<usize> {
+    let mut write_count: usize = 0;
+
+    while sector_count > 0 {
+        if disk_spec.has_parent {
+            return Err(Error::UnsupportedMode);
+        } else {
+            let sector = Sector::new(disk_spec, bat, &sector_index, &sector_count)?;
+
+            let bat_entry = match bat.get(sector.bat_index as usize) {
+                Some(entry) => (*entry).bat_entry,
+                None => {
+                    return Err(Error::InvalidBatIndex);
+                }
+            };
+
+            match bat_entry & vhdx_bat::BAT_STATE_BIT_MASK {
+                vhdx_bat::PAYLOAD_BLOCK_NOT_PRESENT
+                | vhdx_bat::PAYLOAD_BLOCK_UNDEFINED
+                | vhdx_bat::PAYLOAD_BLOCK_UNMAPPED
+                | vhdx_bat::PAYLOAD_BLOCK_UNMAPPED_V095
+                | vhdx_bat::PAYLOAD_BLOCK_ZERO => {
+                    let mut new_size =
+                        round_up!(disk_spec.image_size, vhdx_metadata::BLOCK_SIZE_MIN as u64);
+                    new_size += disk_spec.block_size as u64;
+                    if new_size > size_of::<u64>() as u64 {
```

Comment:
I don't think you meant to write this. This just compares `new_size` to 8 (size of u64)

Reply:
Yes, you are right! Fixed.